### ISSUE TITLE
cosimulation in simulink using FMI toolbox

### ISSUE THREAD

**Issue by yyuanjoy** [lbl-srg/EnergyPlusToFMU#10]
I have created a .idf model and generated .fmu file. I've imported the file to dymola and run it successfully. 
As the next step, I'm trying to import it to simulink using FMI tool box. 
I set sample interval=900, sample offset=0 in FMU block. 
I used variable step solver in simulink and I set the start time=0, and stop time=172800. 

However, I received the following error msg:

module = untitled/FJ1.idf, log level = ERROR: [error][FMU status:Error] fmiDoStep: An error occured in a previous call. First communication time: 900.000000 !=tStart: 0.000000.

I have the .idf, .fmu and .slx files attached. 

Looking forward to your guidance!

Best,
Yuan

[SimulinkFMU.zip](https://github.com/lbl-srg/EnergyPlusToFMU/files/1051634/SimulinkFMU.zip)

**Comment by tsnouidui** [comment]
I don't use the FMI Matlab Toolbox. Hence I don't know what the problem is and can't reproduce it.
The error message seems however to indicate that there was no variables exchanged at t=tStart?

**Comment by yyuanjoy** [comment]
Hi Thierry,

I tried another approach in simulink, which is using the Pilot Support Package (PSP) provided by Matalb, which is a FMI toolbox enabling import of FMI. The PSP toolbaox is free of charge and you can send email to fmi-info@mathworks.com to request it. 

However, I receive another error information as follow when using PSP toolbox. 
[category:ok, status:fmiOK] Command executes to copy content of resources folder: xcopy "D:\YuanYuan\EPlusFMU\simulinktest\slprj\_fmu\a3451a41a6fcda267ae4669549d8e97f\FJ1\resources\resources\"variables.cfg "D:\YuanYuan\EPlusFMU\simulinktest\Output_EPExport_untitled/FMU Co-Simulation\" /Y /I
[category:error, status:fmiError] fmiInstantiateSlave: Could not copy variables.cfg to the output directory folder D:\YuanYuan\EPlusFMU\simulinktest. Instantiation of untitled/FMU Co-Simulation failed.

I checked the resources folder and there is the variables.cfg, but it is under D:\....\resources, instead of D:\....\resources\resources. I'm guess this is a path bug and I hope you could give me some instruction. 

Thank you!

Best,
Yuan

**Comment by tsnouidui** [comment]
Hi Yuan,

I don't know/use this toolbox either but it might be that the PSP toolbox does not pass the correct path to the folder where the FMU has been unzipped to the EnergyPlus FMU. 

I would have expected the path to the resources folder to be D:\YuanYuan\EPlusFMU\simulinktest\slprj_fmu\a3451a41a6fcda267ae4669549d8e97f\FJ1\resources\ without an additional `resources` in the name.

For debugging purposes I suggest to  print out `fmuLocation`  in main.c (https://github.com/lbl-srg/EnergyPlusToFMU/blob/master/SourceCode/EnergyPlus/main.c) and see if it is the folder where the FMU has been unzipped. You can edit the` fmiInstantiateSlave()` function for this purpose.

If `fmuLocation` is not the path to the folder where the FMU has been unzipped, then the FMU will most likely fail since it assumes this to be the case as defined by the specification. 

Thanks!

Thierry

**Comment by yyuanjoy** [comment]
Hi Thierry,

I printed out the fmuLocation, which is file:///D:/YuanYuan/EPlusFMU/PSPtest/slprj/_fmu/70afd75661c4fc1e8a60edfa93a9a7a7/fj1/resources

The unzip location is shown as the graph. Can you identify any issue? If I want to edit thefmiInstantiateSlave() function to make these two location matching, what should I do?

Thank you a lot!!

Best,
Yuan

[cid:image001.png@01D2E463.A0D1B980]

**Comment by tsnouidui** [comment]
This sounds to me like a bug in the toolbox. The fmuLocation shouldn't have any `resources` in its name.
A workaround could be to edit `defines.h` in https://github.com/lbl-srg/EnergyPlusToFMU/blob/master/SourceCode/EnergyPlus/defines.h and  change 

`#define RESOURCES "resources"` to `#define RESOURCES ""`.

This might work. This of course will make the version of E+ you have to no longer conform to the spec. Hence I would rather suggest to contact the tool vendor and see whether they could fix it on their end. If you have the source code of the toolbox you could see if you can not fix it there instead.

**Comment by yyuanjoy** [comment]
Hi Thierry,

As I mentioned, I tried two FMU toolbox in matlab and unfortunately both of them failed. I have send your comments on PSP toolbox to the developer and are still waiting for their response. 

Regrading to the Modelon FMU toolbox, which reported First communication time: 900.000000 !=tStart: 0.000000., here is the response from the developer. 

"The problem seems to be that the EnergyPlus FMU requires the FMU to have a stop time, which we do not set explicitly. The FMI standard allows initializing an FMU with undefined stop time, and we do this and pass 0 as final time, since the parameter should be ignored when the stop time is defined. However, the EnergyPlus FMU you provided still reads the stop time, causing the simulation to fail since the first time step (to t = 900) goes past the final time.

We will look into making things more robust on our end, since Simulink provides a defined stop time that we can use, but unfortunately there is no workaround to use the FMU in Simulink until we have gotten that fix into a release.

I would suggest that you bring this up with the EnergyPlus developers as well, since the FMU does not adhere to the standard."

Is there possibility to solve this issue from your side? Thank you!

Best,
Yuan

**Comment by tsnouidui** [comment]
Yuan,

the specification defines the function `fmiInitializeSlave` as follow

```
fmiStatus fmiInitializeSlave(fmiComponent c, fmiReal tStart,
 fmiBoolean StopTimeDefined, fmiReal tStop);
```
The spec does not explicitely say what the value of tSop needs to be if `StopTimeDefined` is `false`. 
Currently Modelon is setting it to zero which is a value. Another tool could set it to any other value.
Since the specification is ambiguous here, I will add as a requirement for using EnergyPlusToFMU that the stop time must be defined. Otherwise EnergyPlus will have no way to know when the simulation should stop.

**Comment by yyuanjoy** [comment]
Hi Thierry,

Thank you a lot! Please kindly send me the revised code, so that I can try if it works with Modelon FMU toolbox. 

Regarding to the PSP tool box, the developer agreed that there is a bug in their 32 bit toolbox ( the "fmuLocation" difference in v1.0 and v2.0), but they have corrected it the 64 bit toolbox. However, they have given up the update and maintenance on 32 bit toolbox, unless E+ could provide 64 bit toolbox, the reality that 32 bit E+ fmu cannot run with Matlab PSP toolbox is a conclusion. 

Best,
Yuan

**Comment by tsnouidui** [comment]
@yyuanjoy 

I don't think that revising the code to make it not FMI compliant is a good idea. 
If you want to pusue this path, I suggest to checkout the code, do the changes, and use it for your application. 

Best might be at this point to compile E+ as a 64bit FMU and use the 64 bit of the PSP toolbox.

I will close this issue if you don't have any objection.

**Comment by yyuanjoy** [comment]
Hi Thierry,

How can I compile E+ as a 64bit FMU? I simply followed the steps listed on a"application guidance", and it automatically became a 32bit FMU.

Best,
Yuan

**Comment by tsnouidui** [comment]
The latest release is actually setup to export 64 bit FMU using Microsoft Visual Studio 10.0 Professional. If you don't have the Microsoft Visual Studio 10.0 Professional, you will need to adjust the path to the batch files which are in the `Scripts/win` folder of EnergyPlusToFMU so compiler and linker can be found.

**Comment by yyuanjoy** [comment]
Hi Thierry,

I have generated 64bit E+ FMU and make it work in matlab PSP 64bit toolbox. Thank you a lot for the helps!!!!

Please close this  issue.

Best,
Yuan

**Comment by tsnouidui** [comment]
Great.

**Comment by yyuanjoy** [comment]
Hi Thierry,

Just for your information, Modelon has released 2.6.1 beta 1, in which it corrected the stop time bug and now it works well with E+ fmu.

**Comment by tsnouidui** [comment]
Thanks for the info.
